core/lib/mini_merkle_tree/src/lib.rs

```diff
@@ -86,67 +118,135 @@ where
+    /// The range is 0..length, where `0` is the leftmost untrimmed leaf.
+    pub fn merkle_root_and_paths_for_range(&self, length: usize) -> (H256, Vec<H256>, Vec<H256>) {
+        let (mut left_path, mut right_path) = (vec![], vec![]);
+        let root_hash = self.compute_merkle_root_and_path(
+            length - 1,
+            Some(&mut left_path),
+            Some(&mut right_path),
+        );
+        (root_hash, left_path, right_path)
+    }
+
+    /// Adds a raw hash to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push_hash(&mut self, leaf_hash: H256) {
+        self.hashes.push_back(leaf_hash);
+        if self.start_index + self.hashes.len() > self.binary_tree_size {
+            self.binary_tree_size *= 2;
+        }
+    }
+
+    /// Adds a new leaf to the tree (replaces leftmost empty leaf).
+    /// If the tree is full, its size is doubled.
+    /// Note: empty leaves != zero leaves.
+    pub fn push(&mut self, leaf: [u8; LEAF_SIZE]) {
+        let leaf_hash = self.hasher.hash_bytes(&leaf);
+        self.push_hash(leaf_hash);
+    }
+
+    /// Caches the rightmost `count` leaves.
+    /// Does not affect the root hash, but makes it impossible to get the paths to the cached leaves.
+    /// # Panics
+    /// Panics if `count` is greater than the number of non-cached leaves in the tree.
+    pub fn trim_start(&mut self, count: usize) {
+        assert!(self.hashes.len() >= count, "not enough leaves to cache");
+        let mut new_cache = vec![];
+        // Cache is a subset of the path to the first untrimmed leaf.
+        let root = self.compute_merkle_root_and_path(count, None, Some(&mut new_cache));
+        self.hashes.drain(..count);
+        self.start_index += count;
+        // It is important to add the root in case we just trimmed all leaves *and*
+        // the tree will grow on the next push.
+        new_cache.push(root);
+        self.cache = new_cache;
     }
 
     fn compute_merkle_root_and_path(
-        self,
-        mut index: usize,
-        mut merkle_path: Option<&mut Vec<H256>>,
+        &self,
+        mut end_index: usize,
+        mut start_path: Option<&mut Vec<H256>>,
+        mut end_path: Option<&mut Vec<H256>>,
     ) -> H256 {
-        assert!(index < self.hashes.len(), "invalid tree leaf index");
-
         let depth = tree_depth_by_size(self.binary_tree_size);
-        if let Some(merkle_path) = merkle_path.as_deref_mut() {
-            merkle_path.reserve(depth);
+        if let Some(left_path) = start_path.as_deref_mut() {
+            left_path.reserve(depth);
+        }
+        if let Some(right_path) = end_path.as_deref_mut() {
+            right_path.reserve(depth);
         }
 
-        let mut hashes = self.hashes;
-        let mut level_len = hashes.len();
+        let mut hashes = self.hashes.clone();
+        let mut start_index = self.start_index;
```

Comment:
Bikeshedding: This is confusing given that `end_index` is relative and `start_index` is absolute. I'd suggest to call the variable `absolute_start_index`.